src/rewriter/rewrite_proof_rule.h

```diff
@@ -0,0 +1,150 @@
+ *
+ * Rewrite proof rule class
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__REWRITER__REWRITE_PROOF_RULE__H
+#define CVC5__REWRITER__REWRITE_PROOF_RULE__H
+
+#include <string>
+#include <vector>
+
+#include "expr/nary_match_trie.h"
+#include "expr/node.h"
+#include "rewriter/rewrites.h"
+
+namespace cvc5::internal {
+namespace rewriter {
+
+/**
+ * The definition of a (conditional) rewrite rule. An instance of this
+ * class is generated for each DSL rule provided in the rewrite files. The
+ * interface of this class is used by the proof reconstruction algorithm.
+ */
+class RewriteProofRule
+{
+ public:
+  RewriteProofRule();
+  /**
+   * Initialize this rule.
+   * @param id The identifier of this rule
+   * @param userFvs The (user-provided) free variable list of the rule. This
+   * is used only to track the original names of the arguments to the rule.
+   * @param fvs The internal free variable list of the rule. Notice these
+   * variables are normalized such that *all* proof rules use the same
+   * variables, per type. In detail, the n^th argument left-to-right of a given
+   * type T is the same for all rules. This is to facilitate parallel matching.
+   * @param cond The conditions of the rule, normalized to fvs.
+   * @param conc The conclusion of the rule, which is an equality of the form
+   * (= t s), where t is specified as rewriting to s. This equality is
+   * normalized to fvs.
+   * @param isFixedPoint Whether the rule should be applied to fixed point in
+   * the strategy
+   * @param isFlatForm Whether the rule is the flat form of the actual rule
+   * with the given id.
+   */
+  void init(DslPfRule id,
+            const std::vector<Node>& userFvs,
+            const std::vector<Node>& fvs,
+            const std::vector<Node>& cond,
+            Node conc,
+            Node context,
+            bool isFlatForm);
+  /** get id */
+  DslPfRule getId() const;
+  /** get name */
+  const char* getName() const;
+  /** Get user variable list */
+  const std::vector<Node>& getUserVarList() const;
+  /** Get variable list */
+  const std::vector<Node>& getVarList() const;
+  /** The context that the rule is applied in */
+  Node getContext() const { return d_context; }
+  /** Does this rule have conditions? */
+  bool hasConditions() const;
+  /** Get (declared) conditions */
+  const std::vector<Node>& getConditions() const;
+  /**
+   * Get the conditions in context { vs -> ss }. This may involve running the
+   * side conditions of this method.
+   */
+  bool getObligations(const std::vector<Node>& vs,
+                      const std::vector<Node>& ss,
+                      std::vector<Node>& vcs) const;
+  /**
+   * Check match, return true if h matches the head of this rule; notifies
+   * the match notify object ntm.
+   *
+   * Note this method is not run as the main matching algorithm for rewrite
+   * proof reconstruction, which considers all rules in parallel.
```

Comment:
IIUC what you mean is that there is a main algorithm run elsewhere, of which this method is merely a helper?

Reply:
No, its not a helper for the main algorithm.  It would be a helper in a naive, non-parallel version.

Reply:
Could you clarify then this paragraph? I don't get what it's trying to say.